### PR TITLE
Validate jobs in RadixApplication

### PR DESCRIPTION
**Pull request by nilsgstrabo**
- validates RadixApplication configuration

### CODE REVIEW

File: pkg/apis/radixvalidators/validate_ra.go [hunk 1]
```diff
@@ -433,15 +517,31 @@ func environmentHasDynamicTaggingButImageLacksTag(environmentImageTag, component
...
+	if job.SchedulerPort == nil {
+		return SchedulerPortCannotBeEmptyForJobError(job.Name)
+	}
+
+	return nil
+}
+
+func validateJobPayload(job *v1.RadixJobComponent) error {
+	if job.Payload != nil && job.Payload.Path == "" {
+		return SchedulerPortCannotBeEmptyForJobError(job.Name)
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:530]: It should be `PayloadPathCannotBeEmptyForJobError`

File: pkg/apis/radixvalidators/validate_ra.go [hunk 2]
```diff
@@ -85,6 +85,16 @@ func MultipleMatchingPortNamesError(matchingPortName int, publicPortName, compon
 	return fmt.Errorf("There are %d ports with name %s for component %s. Only 1 is allowed", matchingPortName, publicPortName, component)
 }
 
+// SchedulerPortCannotBeEmptyForJobError Scheduler port cannot be empty for job
+func SchedulerPortCannotBeEmptyForJobError(job string) error {
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:89]: Suggestion to use argument name `name` or `jobName` for consistency

File: pkg/apis/radixvalidators/validate_ra.go [hunk 3]
```diff
@@ -85,6 +85,16 @@ func MultipleMatchingPortNamesError(matchingPortName int, publicPortName, compon
...
+	return fmt.Errorf("Scheduler port cannot be empty for %s", job)
+}
+
+// PayloadPathCannotBeEmptyForJobError Payload path cannot be empty for job
+func PayloadPathCannotBeEmptyForJobError(job string) error {
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:94]: Suggestion to use argument name `name` or `jobName` for consistency

File: pkg/apis/radixvalidators/validate_ra.go [hunk 4]
```diff
@@ -422,6 +437,75 @@ func validateComponents(app *radixv1.RadixApplication) []error {
...
+func validateJobComponents(app *radixv1.RadixApplication) []error {
+	errs := []error{}
+	for _, job := range app.Spec.Jobs {
+		if job.Image != "" && (job.SourceFolder != "" || job.DockerfileName != "") {
+			errs = append(errs, PublicImageComponentCannotHaveSourceOrDockerfileSet(job.Name))
+		}
+
+		if usesDynamicTaggingForDeployOnly(job.Image) {
+			if len(job.EnvironmentConfig) == 0 {
+				errs = append(errs, ComponentWithDynamicTagRequiresTagInEnvironmentConfig(job.Name))
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:449]: Suggestion to rename to `ResourceWithDynamicTagRequiresTagInEnvironmentConfig`
and correct its message

File: pkg/apis/radixvalidators/validate_ra.go [hunk 5]
```diff
@@ -422,6 +437,75 @@ func validateComponents(app *radixv1.RadixApplication) []error {
...
+			} else {
+				for _, environment := range job.EnvironmentConfig {
+					if doesEnvExistAndIsMappedToBranch(app, environment.Environment) &&
+						environment.ImageTagName == "" {
+						errs = append(errs,
+							ComponentWithDynamicTagRequiresTagInEnvironmentConfigForEnvironment(job.Name, environment.Environment))
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:455]: Suggestion to rename to `ResourceWithDynamicTagRequiresTagInEnvironmentConfigForEnvironment`
and correct its message

File: pkg/apis/radixvalidators/validate_ra.go [hunk 6]
```diff
@@ -422,6 +437,75 @@ func validateComponents(app *radixv1.RadixApplication) []error {
...
+		if job.Image != "" && (job.SourceFolder != "" || job.DockerfileName != "") {
+			errs = append(errs, PublicImageComponentCannotHaveSourceOrDockerfileSet(job.Name))
+		}
+
+		if usesDynamicTaggingForDeployOnly(job.Image) {
+			if len(job.EnvironmentConfig) == 0 {
+				errs = append(errs, ComponentWithDynamicTagRequiresTagInEnvironmentConfig(job.Name))
+			} else {
+				for _, environment := range job.EnvironmentConfig {
+					if doesEnvExistAndIsMappedToBranch(app, environment.Environment) &&
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:452]: Optional. `doesEnvExistAndIsMappedToBranch` contains weird `if`

File: pkg/apis/radixvalidators/validate_ra.go [hunk 7]
```diff
@@ -422,6 +437,75 @@ func validateComponents(app *radixv1.RadixApplication) []error {
...
+						environment.ImageTagName == "" {
+						errs = append(errs,
+							ComponentWithDynamicTagRequiresTagInEnvironmentConfigForEnvironment(job.Name, environment.Environment))
+					}
+				}
+			}
+
+		}
+
+		err := validateRequiredResourceName("component name", job.Name)
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra.go:462]: "_job_ name"

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 8]
```diff
@@ -290,6 +319,124 @@ func Test_invalid_ra(t *testing.T) {
...
+			ra.Spec.Jobs[0].EnvironmentConfig = []v1.RadixJobComponentEnvironmentConfig{
+				{
+					Environment: noExistingEnvironment,
+				},
+			}
+		}},
+		{"scheduler port is not set", radixvalidators.SchedulerPortCannotBeEmptyForJobError(validRAFirstJobName), func(ra *v1.RadixApplication) {
+			ra.Spec.Jobs[0].SchedulerPort = nil
+		}},
+		{"payload is empty struct", radixvalidators.SchedulerPortCannotBeEmptyForJobError(validRAFirstJobName), func(ra *v1.RadixApplication) {
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra_test.go:354]: `SchedulerPortCannotBeEmptyForJobError` should be `SchedulerPayloadCannotBeEmptyForJobError`

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 9]
```diff
@@ -290,6 +319,124 @@ func Test_invalid_ra(t *testing.T) {
...
+			ra.Spec.Jobs[0].Payload = &v1.RadixJobComponentPayload{}
+		}},
+		{"payload path is empty string", radixvalidators.SchedulerPortCannotBeEmptyForJobError(validRAFirstJobName), func(ra *v1.RadixApplication) {
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra_test.go:357]: `SchedulerPortCannotBeEmptyForJobError` should be `PayloadPathCannotBeEmptyForJobError`

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 10]
```diff
@@ -563,16 +830,30 @@ func Test_PublicPort(t *testing.T) {
 }
 
 func Test_ValidationOfVolumeMounts_Errors(t *testing.T) {
+	type volumeMountsFunc func() []v1.RadixVolumeMount
+	type setVolumeMounts func(*v1.RadixApplication, []v1.RadixVolumeMount)
```
Comment (satr) [inline pkg/apis/radixvalidators/validate_ra_test.go:834]: Suggested name `setVolumeMountsFunc`

**Review by satr — CHANGES_REQUESTED** [review]
Needs couple corrections and some suggestions

File: pkg/apis/radixvalidators/validate_ra.go (hunk 1 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:530]: fixed

File: pkg/apis/radixvalidators/validate_ra.go (hunk 2 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:89]: fixed

File: pkg/apis/radixvalidators/validate_ra.go (hunk 3 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:94]: fixed

File: pkg/apis/radixvalidators/validate_ra.go (hunk 6 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:452]: Is it the line break you refer to?

File: pkg/apis/radixvalidators/validate_ra_test.go (hunk 8 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra_test.go:354]: fixed

File: pkg/apis/radixvalidators/validate_ra_test.go (hunk 9 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra_test.go:357]: fixed

File: pkg/apis/radixvalidators/validate_ra_test.go (hunk 10 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra_test.go:834]: fixed

File: pkg/apis/radixvalidators/validate_ra.go (hunk 7 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:462]: fixed

File: pkg/apis/radixvalidators/validate_ra.go (hunk 4 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:448]: agreed to keep as is

File: pkg/apis/radixvalidators/validate_ra.go (hunk 5 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go:453]: agreed to keep as is

File: pkg/apis/radixvalidators/validate_ra.go (hunk 6 above)
Comment (nilsgstrabo) [reply pkg/apis/radixvalidators/validate_ra.go]: agreed to keep as is

**Review by satr — APPROVED** [review]

**Pull request merged**